components/scream/src/dynamics/homme/atmosphere_dynamics_rayleigh_friction.cpp

```diff
@@ -0,0 +1,83 @@
+#include "atmosphere_dynamics.hpp"
+
+// Scream includes
+#include "share/util/scream_common_physics_functions.hpp"
+
+// Homme includes
+#include "dynamics/homme/homme_dimensions.hpp"
+
+namespace scream {
+
+void HommeDynamics::rayleigh_friction_init()
+{
+  constexpr int N = HOMMEXX_PACK_SIZE;
+  using KT = KokkosTypes<DefaultDevice>;
+  using Pack = ekat::Pack<Real,N>;
+
+  const int nlevs = m_dyn_grid->get_num_vertical_levels();
+  const auto npacks= ekat::PackInfo<Pack::n>::num_packs(nlevs);
+  m_otau = decltype(m_otau)("otau", npacks);
+
+  // Rayleigh friction paramaters
+  m_rayk0     = m_params.get<int>("Rayleigh Friction Vertical Level", 1);
+  m_raykrange = m_params.get<Real>("Rayleigh Friction Range", 0.0);
+  m_raytau0   = m_params.get<Real>("Rayleigh Friction Decay Time", 5.0);
+
+  // Calculate decay rate profile, otau.
+  // If m_raytau0 == 0 no Rayleigh friction is applied.
+  if (m_raytau0 != 0) {
+    Real krange; // range of rayleigh friction profile
+    Real tau0;   // approximate value of decay time at model top
+    Real otau0;  // inverse of tau0
+
+    krange = m_raykrange;
+    if (m_raykrange == 0) krange = m_rayk0/2.0;
+
+    tau0 = 86400.0*m_raytau0; // convert to seconds
+    otau0 = 0;
+    if (tau0 != 0) otau0 = 1.0/tau0;
+
+    auto otau = m_otau;
+    const auto nlevs = m_phys_grid->get_num_vertical_levels();
+    const int npacks = ekat::PackInfo<Pack::n>::num_packs(nlevs);
+
+    Kokkos::parallel_for(KT::RangePolicy(0, npacks),
+                         KOKKOS_LAMBDA (const int ilev) {
+      const auto range_pack = ekat::range<Pack>(ilev*Pack::n);
+      const Pack x = (m_rayk0 - range_pack)/krange;
+      otau(ilev) = otau0*(1.0 + ekat::tanh(x))/2.0;
+    });
+  }
+}
+
+void HommeDynamics::rayleigh_friction_apply(const Real dt) const
+{
+  constexpr int N = HOMMEXX_PACK_SIZE;
+  using PF = PhysicsFunctions<DefaultDevice>;
+  using KT = KokkosTypes<DefaultDevice>;
+  using Pack = ekat::Pack<Real,N>;
+  using ESU = ekat::ExeSpaceUtils<KT::ExeSpace>;
+
+  const auto ncols = m_phys_grid->get_num_local_dofs();
+  const auto nlevs = m_phys_grid->get_num_vertical_levels();
+  const auto npacks= ekat::PackInfo<N>::num_packs(nlevs);
+
+  const auto horiz_winds_view = get_field_out("horiz_winds").get_view<Pack***>();
+  const auto T_mid_view       = get_field_out("T_mid").get_view<Pack**>();
+
+  const auto policy = ESU::get_default_team_policy(ncols, npacks);
+  Kokkos::parallel_for(policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const int& icol = team.league_rank();
+
+    // If m_raytau0 == 0 no Rayleigh friction is applied.
+    if (m_raytau0 != 0) {
```

Comment:
m_raytau0 is a class var; this can crash on the GPU. Copy to a local for lambda capture.